Updatecli && Maven (#495)
updatecli/website: content/en/blog/2022/09/updatecli_java.adoc

---
title: "Updatecli with Java"
description: "Automate Maven Dependencies"
lead: "Let's automate Maven dependencies"
date: 2022-09-16T09:19:42+01:00
draft: false
weight: 50
images: [""]
contributors: ["olblak"]
---
:toc: right
:hide-uri-scheme:

= Updatecli && Java? 

I recently spent some time on a Java application using Maven and wondered what if I could bump all my Java dependencies in one command, directly from my machine?

How much time would it save me? 

A bit of context, when I started working on Updatecli I wanted a tool that would work from anywhere and allow me to automate file changes, stored on git repositories.  

I could have leveraged the Updatecli XML resource (https://www.updatecli.io/docs/plugins/resource/xml[more here]), to automate my pom.xml. But considering that pom.xml files are well structured, I was more in favour of leveraging some work I did on a new Updatecli feature named “autodiscovery”.

> So instead of relying on an updatecli manifest that describes how a pom.xml should be updated,  “autodiscovery” allows Updatecli to analyse directories, files, to automate common pattern.

Reviewer: ```suggestion
So instead of relying on an updatecli manifest that describes how a pom.xml should be updated,  “autodiscovery” allows Updatecli to analyse directories, and files to automate common manifest patterns.
```

Author: that's a quick review :D

---
image::/images/blog/2022/09/maven.png["Java"]
---

Before sharing my thoughts, I propose to let you build your own opinion, first.

Here is how you would do

1. Install Updatecli

> 2. Run Updatecli in your project directory

Reviewer: run Updatecli "autodiscovery" and/or an updatecli discovery command snippet?

3. Reach your conclusion

== Install

First we need to install Updatecli.

> Updatecli is a Golang application available for Linux/Windows/MacOSx. All you have to do is download it and execute it.

Reviewer: ```suggestion
As Updatecli is a Golang application available for Linux/Windows/MacOSx, all you have to do is download and execute it.
```

For Homebrew users:

* `brew tap updatecli/updatecli`
* `brew install updatecli`

> Or download the right package for your favourite OS/Architecture on 

Reviewer: ```suggestion
Or download the right package for your favourite OS/Architecture from <link below>
```

=> https://github.com/updatecli/updatecli/releases/latest

== Run

> Updatecli works in two different ways.

Reviewer: ```suggestion
Updatecli works in two different modes:
```
(Nit)

Author: I prefer your suggestion

Either it detects automatically what needs to be changed using the experimental “autodiscovery” feature.

Or in a more declarative way, **you** specify **your** update scenario using a YAML manifest.

Before running any command, we need a Java project using Maven. 

Use whatever you want, but I’ll use the Jenkins repository as it contains many dependencies. 

```
cd /tmp

> git clone [git@github.com](mailto:git@github.com):jenkinsci/jenkins.git

Author: ```suggestion
git clone https://github.com/jenkinsci/jenkins.git
```

cd jenkins
```

=== Autodiscovery

“Autodiscovery” is the easiest approach, all you have to do, is going at the root of your project directory and run:

> * `updatecli diff –local-autodiscovery –experimental`: to see what would change

Author: ```suggestion
* `updatecli diff --local-autodiscovery --experimental`: to see what would change
```

> * `updatecli apply –local-autodiscovery –experimental`: to apply what you saw in the previous command output

Author: ```suggestion
* `updatecli apply --local-autodiscovery --experimental`: to apply what you saw in the previous command output
```

By default the autodiscovery doesn’t handle any git service operation. This could be provided using a manifest that requires credentials.

.updatecli.d/manifest.yaml
```
scms:
  default:
	kind: github
	spec:
  	    branch: main
  	    owner: <your owner repository>
  	    repository: <your repository name>
  	    token: <your token>
  	    username: "your username"
pullrequests:
  default:
	kind: github
	spec:

autodiscovery:
  scmid: default
  crawlers:
	maven:
  	    enable: true
``` 

> * `updatecli diff –local-autodiscovery –experimental –config updatecli.d/manifest.yaml`: to see what would change

Author: ```suggestion
* `updatecli diff --local-autodiscovery --experimental --config updatecli.d/manifest.yaml`: to see what would change
```

> * `updatecli apply –local-autodiscovery –experimental –config updatecli.d/manifest.yaml`: to apply what you saw in the previous command output

Author: ```suggestion
* `updatecli apply --local-autodiscovery --experimental --config updatecli.d/manifest.yaml`: to apply what you saw in the previous command output
```

Tips: If you use an IDE that supports Jsonschema by default, such as vscode, then you can benefit from auto-completion by creating the updatecli manifest inside the directory named “updatecli.d”. While the directory “updatecli.d” is not mandatory, It allows IDE that support “json schema store” to know which schema to use, to benefit from autocompletion such as IntelliJ or VsCode

Once again you can use the following command:

> * `updatecli diff –local-autodiscovery –experimental –config updatecli.d/maven.yaml`: to see what would change

Author: ```suggestion
* `updatecli diff --local-autodiscovery --experimental --config updatecli.d/maven.yaml`: to see what would change
```

> * `updatecli apply –local-autodiscovery –experimental  –config updatecli.d/maven.yaml`: to apply what you saw in the previous command output

Author: ```suggestion
* `updatecli apply --local-autodiscovery --experimental  --config updatecli.d/maven.yaml`: to apply what you saw in the previous command output
```
